elyra/pipeline/component.py

```diff
@@ -275,73 +236,158 @@ class ComponentReader(LoggingConfigurable):
+        """
+        Returns a list of absolute paths to component specification file(s)
+        based on the array of potentially relative location given
+        """
         raise NotImplementedError()
 
 
 class FilesystemComponentReader(ComponentReader):
     """
-    Read a component definition from the local filesystem
+    Read a singular component definition from the local filesystem
     """
     type = 'filename'
 
-    def read_component_definition(self, registry_entry: SimpleNamespace) -> Optional[str]:
-        component_path = os.path.join(ENV_JUPYTER_PATH[0], 'components', registry_entry.location)
-        if not os.path.exists(component_path):
-            self.log.warning(f"Invalid location for component: {registry_entry.id} -> {component_path}")
+    def determine_location(self, location_path: str) -> str:
+        """
+        Determines the absolute location of a given path. Error
+        checking is delegated to the calling function
+        """
+        # Expand path to include user home if necessary
+        path = os.path.expanduser(location_path)
+
+        # Check for absolute path
+        if os.path.isabs(path):
+            return path
+
+        # If path is not absolute, default to the Jupyter share location
+        path = os.path.join(ENV_JUPYTER_PATH[0], 'components', path)
+        return path
+
+    def read_component_definition(self, location: str) -> Optional[str]:
+        if not os.path.exists(location):
+            self.log.warning(f"Invalid location for component: {location}")
             return None
 
-        with open(component_path, 'r') as f:
+        with open(location, 'r') as f:
             return f.read()
 
+    def get_list_of_paths(self, paths: List[str], parser_file_types: List[str]) -> List[str]:
+        absolute_paths = []
+        for path in paths:
+            absolute_path = self.determine_location(path)
+            if not os.path.exists(absolute_path):
+                self.log.warning(f"File does not exist -> {absolute_path}")
+            absolute_paths.append(absolute_path)
+        return absolute_paths
+
+
+class DirectoryComponentReader(FilesystemComponentReader):
+    """
+    Read component definitions from a local directory
+    """
+    type = 'directory'
+
+    def get_list_of_paths(self, paths: List[str], parser_file_types: List[str]) -> List[str]:
+        absolute_paths = []
+        for path in paths:
+            absolute_path = self.determine_location(path)
+            if not os.path.exists(absolute_path):
+                self.log.warning(f"Invalid directory -> {absolute_path}")
+                continue
+
+            for filename in os.listdir(absolute_path):
+                if filename.endswith(tuple(parser_file_types)):
+                    absolute_paths.append(os.path.join(absolute_path, filename))
+
+        return absolute_paths
+
+    @property
+    def base_type(self):
+        return super().type
```

Comment:
Hmm - so this is returning `filename` - is that correct?  We might want a comment about that.

Reply:
Good point. Assuming my comment above explaining the difference between `type` and `base_type` doesn't change the way we want to do things, I'll definitely add that!